dds/idl/be_global.cpp

```diff
@@ -646,15 +646,22 @@ bool BE_GlobalData::is_topic_type(AST_Decl* node)
 {
   TopicAnnotation* topic = dynamic_cast<TopicAnnotation*>(
     builtin_annotations_["::@topic"]);
-  return topic->node_value(node) || !is_nested(node);
+  TopicValue value;
+  return topic->node_value_exists(node, value);
 }
 
 bool BE_GlobalData::is_nested(AST_Decl* node)
 {
+  // @topic overrides @nested and @default_nested.
+  if (is_topic_type(node)) {
+    return false;
+  }
```

Comment:
You should revert these lines. `is_topic_type` is being called from other places in the compiler. The naming is probably a bit poor, but `is_nested` was just meant to be a helper to `is_topic_type`. If you need to know if something is "nested" you can just invert the result of `is_topic_type`.

Reply:
I disagree.  According to the spec `@topic` overrides `@nested` and `@default_nested`.  I could define a non-topic type with `@nested`.  I don't know if that is relevant or meaningful, but it does seem to be legal.

Reply:
>  I could define a non-topic type with @nested.

That's just one of the definitions of a non-topic type. I think the spec handles this very poorly. These are all sides of the same coin, which is if a type is a topic type. By doing this you're breaking the existing functionality.

Reply:
See corrected comment above.
You can have (1) nested non-topic types, (2) non-nested topic types, and (3) non-nested non-topic types.
How are we supposed to handle (3)?
If we need to handle (3) and want to preserve existing functionality, then we need a compelling reason that is well documented and communicated to the user.  Otherwise, I think we should implement the spec correctly and change the code accordingly.

Reply:
I'm assuming you mean by these cases you mean annotations since there is no such thing as a "non-nested non-topic type".

A type with no `@nested` annotation or `@topic` defaults to the "default nested" property of the module which can go up to the root module's "default nested" property which can be set by the user using a command line argument.

I don't know what you mean by the last paragraph but this is all documented in the devguide in section 2.1.1.4
"Topic Types vs. Nested Types".

Reply:
I'm assuming that something in the spec says that `@nested(FALSE)` is equivalent to `@topic`.  Where is that located?  If its not in there, then why did we make that assumption?

Reply:
So @oschwaldp-oci pointed me to 7.2.2.5.  I think what the spec writers intended was that type annotated with `@nested(TRUE)` are optimized because they must be nested, type annotated with `@nested(FALSE)` are candidates for being topics, and types annotated with `@topic` should definitely be treated as topics with the name and platform and such.  Thus, treating `@nested(FALSE)` as `@topic` will work but it may be more than the spec writers intended.

I'll correct `is_topic_type` and `is_nested`.

Reply:
I honestly don't think was that thought out. `@topic` was only added after I started working on the annotations. Adam filed an issue because the spec assumed topic types were the default and it was obvious it would be helpful to have something like `@topic` if we were keeping the original behavior in OpenDDS.